src/Symfony/Component/Messenger/Handler/HandlersLocator.php

```diff
@@ -33,9 +33,19 @@ public function __construct(array $handlers)
     }
 
     /**
-     * {@inheritdoc}
+     * @internal
      */
-    public function getHandlers(Envelope $envelope): iterable
+    public static function listTypes(Envelope $envelope): array
+    {
+        $class = \get_class($envelope->getMessage());
+
+        return [$class => $class]
+            + class_parents($class)
+            + class_implements($class)
+            + ['*' => '*'];
```

Comment:
This seems unecessary in the context of an exactly-one-handler bus, as `[$class => $class]` would be enough.

In order to detect the issues ASAP, type `*` could be prohibited in the `AtLeastOneHandlerLocator` constructor.